scanpy/preprocessing/simple.py

```diff
@@ -122,38 +123,75 @@ def filter_cells(data, min_counts=None, min_genes=None, max_counts=None,
     return cell_subset, number_per_cell
 
 
-def filter_genes(data, min_cells=None, min_counts=None, copy=False):
-    """Filter genes based on minimal number of cells or counts.
+def filter_genes(data, min_cells=None, min_counts=None, max_counts=None,
+                 max_cells=None, copy=False):
+    """Filter genes based on number of cells or counts.
 
     Keep genes that have at least `min_counts` counts or are expressed in at
-    least `min_cells` cells.
+    least `min_cells` cells.  Or keep genes that have at most `max_counts`
+    counts or are expressed in at most `max_cells` cells.
+
+    Only provide one of the optional arguments (`min_counts`, `min_cells`,
+    `max_counts`, `max_cells`) per call.
 
     See :func:`~scanpy.api.pp.filter_cells`.
     """
-    if min_cells is not None and min_counts is not None:
-        raise ValueError('Either specify min_counts or min_cells, but not both.')
-    if min_cells is None and min_counts is None:
-        raise ValueError('Provide one of min_counts or min_cells.')
+    multi_option_error = False
+    
+    if min_cells is not None:
+        if min_counts is not None or max_cells is not None or max_counts is not None:
+            multi_option_error = True
+    if min_counts is not None:
+        if min_cells is not None or max_cells is not None or max_counts is not None:
+            multi_option_error = True
+    if max_cells is not None:
+        if min_cells is not None or min_counts is not None or max_counts is not None:
+            multi_option_error = True
+    if max_counts is not None:
+        if min_cells is not None or min_counts is not None or max_cells is not None:
+            multi_option_error = True
+    if multi_option_error:
+        raise ValueError("Only provide one of the optional arguments (`min_counts`," +
+                         "`min_cells`, `max_counts`, `max_cells`) per call.")
```

Comment:
This is confusing and super long. I wouldn’t see at a glance what it really does, but it looks like it’s just doing this:

```py
n_given_options = sum(option is not None for option in [min_cells, min_counts, max_cells, max_counts])
if n_given_options > 1:
    raise ValueError(...)
```

Is that correct?

We should also maybe add a check if they all are `None`, like `if n_given_options == 0`

Reply:
That is better, yes.  I originally started doing the conditional check like in the [filter_cells() method](https://github.com/theislab/scanpy/blob/master/scanpy/preprocessing/simple.py#L87), since it had all the options, but realized that checking would fail if someone passed some options together, such as min_counts and max_genes.  I will admit to often not doing the (quite pythonic) inline loop/conditional as you have done, but in this case it reads well.

Reply:
Please also remember to see my note that this isn't actually working still.  Not the option parsing, but my attempt to make it filter properly fails with a "ValueError: Length of values does not match length of index".